### PR TITLE
Add api endpoint for rapidpro optout hooks

### PR DESCRIPTION
**Pull request by rudigiesler**
When a person opts out from a rapidpro flow, we want to put them back on the old messageset that they were on before.

To do that, we look for the last migration that the identity had, and deactivate any subscriptions to the migrated to messageset that the identity might have.

We then create a new subscription to the migrated from messageset, using the mapper class to map the sequence number.

### CODE REVIEW

File: mapper/api_views.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        try:
+            seed_uuid = UUID(contact.fields[settings.RAPIDPRO_UUID_FIELD])
+        except (ValueError, TypeError, KeyError):
+            raise InvalidRapidproContact(
+                'Rapidpro contact {} has an invalid {} field.'.format(
+                    contact.uuid, settings.RAPIDPRO_UUID_FIELD))
+
+        migration = self.find_migration(seed_uuid)
+        existing_subs = self.get_existing_subscriptions(
+            seed_uuid, migration.to_messageset)
```
Comment (KaitCrawford) [inline mapper/api_views.py]: what if the gates subscriptions have multiple messagesets? eg a `gates_prebirth` that has a `gates_postbirth` as the `next_messageset`? Then the migration may point to the first messageset but the user may have moved onto the second messageset.
I _think_ this is a possible design for some of the investigations

File: mapper/api_views.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+        sequence_number = map_sequence(
+            self.get_messageset(migration.to_messageset)['short_name'],
+            self.get_messageset(migration.from_messageset)['short_name'],
+            old_sub['next_sequence_number']
+        )
+
+        new_sub = self.sbm_client.create_subscription({
+            'identity': seed_uuid,
+            'messageset': migration.from_messageset,
```
Comment (KaitCrawford) [inline mapper/api_views.py]: This will have the same problem if Gates subscriptions straddle EDDs

File: mapper/test_utils.py [hunk 3]
```diff
@@ -52,10 +54,28 @@ def mock_update_subscription(subscription_id):
 
 
 def mock_create_subscription():
-    responses.add(
+    def mirror_callback(request):
+        return (200, {}, request.body)
```
Comment (KaitCrawford) [inline mapper/test_utils.py]: I think the api returns 201 on successful creations

**Comment by alexmuller** [comment]
I think this makes sense to me but it'd be good for somebody else to review as it is pretty complex. Just a couple of comments:

- The automatic "Created at" field on the 2 new models is nice (I couldn't remember if Django did that automatically but it looks like it does)
- Instead of a `RapidproOptout` I'd be tempted to call it a `GatesOptout` or a `MessagingExperimentOptout` or something like that. I guess technically the user is opting out of using RapidPro though (even if they don't know it) so I'm not too bothered by this.

**Review by KaitCrawford — COMMENTED** [review]
Left some comments

**Comment by rudigiesler** [comment]
@alexmuller 

- Yup, the created at field is automatic
- I called it a rapidpro optout because the logic is specific to rapidpro. It does a reverse lookup back to rapidpro to get the full contact details, to get the seed identity uuid, and then goes forward with that. Any other service would need different logic, and therefore a different endpoint.

File: mapper/api_views.py (hunk 1 above)
Comment (rudigiesler) [reply mapper/api_views.py]: Yup, this is something I hadn't thought of. Let me chat to some people and get back to you on whether this is a valid use case.

**Comment by alexmuller** [comment]
> I called it a rapidpro optout because the logic is specific to rapidpro. It does a reverse lookup back to rapidpro to get the full contact details

Makes sense, thanks!

**Review by KaitCrawford — APPROVED** [review]
:+1:

**Pull request merged**